runtime/near-vm-runner/src/cache.rs

```diff
@@ -1,37 +1,148 @@
+use near_primitives::hash::CryptoHash;
+use near_vm_errors::CacheError::{DeserializationError, SerializationError, WriteError};
 use near_vm_errors::VMError;
-use near_vm_logic::VMConfig;
-
-/// Cache size in number of cached modules to hold.
-#[cfg(not(feature = "no_cache"))]
-const CACHE_SIZE: usize = 128;
-// TODO: store a larger on-disk cache
-
-#[cfg(not(feature = "no_cache"))]
-cached_key! {
-    MODULES: SizedCache<(Vec<u8>, u64), Result<wasmer_runtime::Module, VMError>>
-        = SizedCache::with_size(CACHE_SIZE);
-    Key = {
-        (code_hash, config.non_crypto_hash())
-    };
-
-    fn compile_module(code_hash: Vec<u8>, code: &[u8], config: &VMConfig
-        ) -> Result<wasmer_runtime::Module, VMError> = {
-        let prepared_code = prepare::prepare_contract(code, config)?;
-        wasmer_runtime::compile(&prepared_code).map_err(|err| err.into_vm_error())
-    }
-}
+use near_vm_logic::{VMConfig, VMKind};
+use std::convert::TryFrom;
+use std::fmt;
+use wasmer_runtime::{compiler_for_backend, Backend};
+use wasmer_runtime_core::cache::Artifact;
+use wasmer_runtime_core::load_cache_with;
 
-#[cfg(feature = "no_cache")]
 pub(crate) fn compile_module(
-    _code_hash: Vec<u8>,
     code: &[u8],
     config: &VMConfig,
 ) -> Result<wasmer_runtime::Module, VMError> {
     let prepared_code = prepare::prepare_contract(code, config)?;
     wasmer_runtime::compile(&prepared_code).map_err(|err| err.into_vm_error())
 }
+
+#[derive(Debug, Clone, BorshDeserialize, BorshSerialize)]
+enum ContractCacheKey {
+    Version1 { code_hash: CryptoHash, vm_config_non_crypto_hash: u64, vm_kind: u32 },
+}
+
+#[derive(Debug, Clone, BorshDeserialize, BorshSerialize)]
+enum CacheRecord {
+    Error(VMError),
+    Code(Vec<u8>),
+}
+
+fn get_key(code_hash: &[u8], code: &[u8], vm_kind: VMKind, config: &VMConfig) -> CryptoHash {
+    let hash = match CryptoHash::try_from(code_hash) {
+        Ok(hash) => hash,
+        // Sometimes caller doesn't compute code_hash, so hash the code ourselves.
+        Err(_e) => near_primitives::hash::hash(code),
+    };
+    let key = ContractCacheKey::Version1 {
+        code_hash: hash,
+        vm_config_non_crypto_hash: config.non_crypto_hash(),
+        vm_kind: vm_kind as u32,
+    };
+    near_primitives::hash::hash(&key.try_to_vec().unwrap())
+}
+
+fn cache_error(error: VMError, key: &CryptoHash, cache: &dyn CompiledContractCache) -> VMError {
+    let record = CacheRecord::Error(error.clone());
+    if cache.put(&(key.0).0, &record.try_to_vec().unwrap()).is_err() {
+        // That's fine, just cannot cache compilation error.
+        println!("Cannot cache an error");
+    }
+    error
+}
+
+fn compile_and_serialize_wasmer(
+    wasm_code: &[u8],
+    config: &VMConfig,
+    key: &CryptoHash,
+    cache: &dyn CompiledContractCache,
+) -> Result<wasmer_runtime::Module, VMError> {
+    let module = compile_module(wasm_code, config).map_err(|e| cache_error(e, &key, cache))?;
```

Comment:
Let's remove `cache_error` method and replace this into something like this:

```suggestion
    let module = compile_module(wasm_code, config).map_err(|compilation_error| {
      let record = CacheRecord::Error(compilation_error.clone);
      if let Err(_cache_err) = cache.put(&(key.0).0, &record.try_to_vec().unwrap()) {
        // TODO: Change to debug or warn
        println!("Cannot cache an error");
        VMError::CacheError(WriteError { key} )
      } else {
        compilation_error
      }
    })?;
```

Reply:
IMO having named method performing operation is usually better.